contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala

```diff
@@ -0,0 +1,324 @@
+  }
+
+  override def write(
+      path: Path,
+      actions: Iterator[String],
+      overwrite: Boolean,
+      hadoopConf: Configuration
+  ): Unit = {
+    val (fs, resolvedPath) = resolvePath(path, hadoopConf)
+
+    if (overwrite) {
+      return writeActions(fs, path, actions);
+    };
+
+    val tablePath = getTablePath(resolvedPath)
+
+    if (FileNames.isDeltaFile(path)) {
+      val version = FileNames.deltaVersion(path)
+
+      if (version > 0) {
+        val prevVersion = version - 1
+        val prevPath = FileNames.deltaFile(tablePath, prevVersion)
+        getExternalEntry(tablePath, prevPath) match {
+          case Some(entry) => fixDeltaLog(fs, entry)
+          case None =>
+            if (!fs.exists(prevPath)) {
+              throw new java.nio.file.FileSystemException(
+                s"previous commit ${prevPath} doesn't exist"
+              )
+            }
+          // previous commit exists in fs but not in dynamodb
+        }
+      } else {
+        getExternalEntry(tablePath, path) match {
+          case Some(entry) =>
+            if (entry.complete && !fs.exists(path)) {
+              throw new java.nio.file.FileSystemException(
+                s"Old entries for ${tablePath} still exist in the database"
+              )
+            }
+          case None => ;
+        }
+      }
+
+    }
+
+    val tempPath = createTemporaryPath(resolvedPath)
+
+    val entry =
+      ExternalCommitEntry(
+        tablePath,
+        resolvedPath.getName,
+        tempPath,
+        false,
+        None
+      )
+
+    logDebug(s"write: writing to temporary file ${tempPath}")
+    writeActions(fs, entry.absoluteTempPath(), actions);
+
+    // commit to dynamodb E(N, false) and exclude commitTime
+    logDebug(s"write incomplete entry to dynamodb ${entry}")
+    putExternalEntry(entry, overwrite = false)
+
+    try {
+      writeCopyTempFile(fs, entry.absoluteTempPath(), resolvedPath)
+      // Commit (with overwrite=true) to DynamoDB entry E(N, true)
+      // with commitTime attribute in epoch seconds
+      writePutCompleteDbEntry(entry)
+    } catch {
+      case e: Throwable =>
+        logInfo(s"${e.getClass.getName}: ignoring recoverable error: $e")
+    }
+  }
+
+  /*
+   * Write to db in exclusive way.
+   * Method should throw @java.nio.file.FileAlreadyExistsException if path exists in cache.
+   */
+  protected def putExternalEntry(
+      ExternalCommitEntry: ExternalCommitEntry,
+      overwrite: Boolean = false
+  ): Unit
+
+  protected def getExternalEntry(
+      tablePath: Path,
+      jsonPath: Path
+  ): Option[ExternalCommitEntry]
+
+  protected def getLatestExternalEntry(
+      tablePath: Path
+  ): Option[ExternalCommitEntry]
+
+  override def isPartialWriteVisible(path: Path): Boolean = false
+}
+
+/**
+ * The file metadata to be stored in the external db
+ * @param tablePath path to the root of the table (path/to/my/table)
+ * @param tempPath relative path to the `tablePath`
```

Comment:
as you commented, make this relative to _delta_log

https://github.com/delta-io/delta/pull/339#discussion_r817596670

Reply:
and document that is it so

Reply:
@mrk-its